drivers/sensor/icm20948/icm20948.c

```diff
@@ -0,0 +1,581 @@
+/*
+ * Copyright (c) 2023, Lazaro O'Farrill
+ * email: lazaroofarrill@gmail.com
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+
+#define DT_DRV_COMPAT invensense_icm20948
+
+#include "zephyr/kernel.h"
+#include "zephyr/sys/util.h"
+#include "stdint.h"
+#include "zephyr/drivers/i2c.h"
+#include "zephyr/drivers/sensor.h"
+#include "zephyr/init.h"
+#include "zephyr/logging/log.h"
+#include "zephyr/sys/__assert.h"
+#include "icm20948.h"
+
+/* Created by Lazaro O'Farrill on 09/03/2023. */
+
+LOG_MODULE_REGISTER(ICM20948, CONFIG_SENSOR_LOG_LEVEL);
+
+static int icm20948_bank_select(const struct device *dev, uint8_t bank)
+{
+	if (bank > 3) {
+		return -EINVAL;
+	}
+	const struct icm20948_config *cfg = dev->config;
+	const uint8_t reg_bank_mask = 0x30;
+
+	int ret =
+		i2c_reg_update_byte_dt(&cfg->i2c, ICM20948_REG_BANK_SEL, reg_bank_mask, bank << 4);
+
+	if (ret) {
+		LOG_ERR("Error switching banks");
+		return ret;
+	}
+
+	return 0;
+}
+
+static void icm20948_convert_temp(struct sensor_value *val, int16_t raw_val)
+{
+
+	/* Offset by 21 degrees Celsius */
+	int64_t in100 =
+		(raw_val * 100) + (ICM20948_ROOM_TEMP_OFFSET_DEG * ICM20948_TEMP_SENSITIVITY_X100);
+
+	/* Whole celsius */
+	val->val1 = in100 / ICM20948_TEMP_SENSITIVITY_X100;
+
+	/* Micro celsius */
+	val->val2 = ((in100 - (val->val1) * ICM20948_TEMP_SENSITIVITY_X100) * INT64_C(1000000)) /
+		    ICM20948_TEMP_SENSITIVITY_X100;
+}
+
+static int icm20948_convert_accel(const struct icm20948_config *cfg, int32_t raw_accel_value,
+				  struct sensor_value *output_value)
+{
+	int64_t sensitivity = 0; /* value equivalent for 1g */
+
+	switch (cfg->accel_fs) {
+	case ACCEL_FS_SEL_2G:
+		sensitivity = 16384;
+		break;
+	case ACCEL_FS_SEL_4G:
+		sensitivity = 8192;
+		break;
+	case ACCEL_FS_SEL_8G:
+		sensitivity = 4096;
+		break;
+	case ACCEL_FS_SEL_16G:
+		sensitivity = 2048;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Convert to micrometers/s^2 */
+	int64_t in_ms = raw_accel_value * SENSOR_G;
+
+	/* meters/s^2 whole values */
+	output_value->val1 = in_ms / (sensitivity * 1000000LL);
+
+	/* micrometers/s^2 */
+	output_value->val2 = (in_ms - (output_value->val1 * sensitivity * 1000000LL)) / sensitivity;
+
+	return 0;
+}
+
+static int icm20948_convert_gyro(struct sensor_value *val, int16_t raw_val, uint8_t gyro_fs)
+{
+	int64_t gyro_sensitivity_10x = 0; /* value equivalent for 10x gyro reading deg/s */
+
+	switch (gyro_fs) {
+	case GYRO_FS_250:
+		gyro_sensitivity_10x = 1310;
+		break;
+	case GYRO_FS_500:
+		gyro_sensitivity_10x = 655;
+		break;
+	case GYRO_FS_1000:
+		gyro_sensitivity_10x = 328;
+		break;
+	case GYRO_FS_2000:
+		gyro_sensitivity_10x = 164;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	int64_t in10_rads = (int64_t)raw_val * SENSOR_PI * 10LL;
+
+	/* Whole rad/s */
+	val->val1 = (int32_t)(in10_rads / (gyro_sensitivity_10x * 180LL * 1000000LL));
```

Comment:
Can you put a define for the magic number?

Reply:
done

Reply:
https://docs.zephyrproject.org/apidoc/latest/group__sensor__interface.html#ga39d4b84f5d792e27b2d6d5eb6a2ccb0d.  There is actually a function  within sensor api in order to convert degree to rad. Could that not be used?